src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.ConnectionPools.cs

```diff
@@ -13,68 +13,74 @@
+
             private readonly ConnectionFactory _connectionFactory;
             private readonly ReaderWriterLockSlim _shutdownLock;
-            private readonly int _maxPoolConnections;
+            private readonly int _capacityPerService;
 
             // keyed to serviceName. each connection is for specific service such as CodeAnalysisService
-            private readonly ConcurrentDictionary<RemoteServiceName, ConcurrentQueue<Connection>> _pools;
+            private readonly ConcurrentDictionary<RemoteServiceName, Pool> _pools;
 
             private bool _isDisposed;
 
-            public ConnectionPool(ConnectionFactory connectionFactory, int maxPoolConnection)
+            public ConnectionPools(ConnectionFactory connectionFactory, int capacity)
             {
                 _connectionFactory = connectionFactory;
-                _maxPoolConnections = maxPoolConnection;
+                _capacityPerService = capacity;
 
                 // initial value 4 is chosen to stop concurrent dictionary creating too many locks.
                 // and big enough for all our services such as codeanalysis, remotehost, snapshot and etc services
-                _pools = new ConcurrentDictionary<RemoteServiceName, ConcurrentQueue<Connection>>(concurrencyLevel: 4, capacity: 4);
+                _pools = new ConcurrentDictionary<RemoteServiceName, Pool>(concurrencyLevel: 4, capacity: 4);
 
                 _shutdownLock = new ReaderWriterLockSlim(LockRecursionPolicy.NoRecursion);
             }
 
-            public async Task<Connection> GetOrCreateConnectionAsync(RemoteServiceName serviceName, CancellationToken cancellationToken)
+            public async Task<RemoteServiceConnection> GetOrCreateConnectionAsync(RemoteServiceName serviceName, CancellationToken cancellationToken)
             {
-                var queue = _pools.GetOrAdd(serviceName, _ => new ConcurrentQueue<Connection>());
-                if (queue.TryDequeue(out var connection))
+                var pool = _pools.GetOrAdd(serviceName, _ => new Pool(this));
+                if (pool.TryAcquire(out var connection))
                 {
-                    return new PooledConnection(this, serviceName, connection);
+                    return connection;
                 }
 
-                var newConnection = await _connectionFactory(serviceName, cancellationToken).ConfigureAwait(false);
-                return new PooledConnection(this, serviceName, newConnection);
+                return await _connectionFactory(serviceName, pool, cancellationToken).ConfigureAwait(false);
             }
 
-            private void Free(RemoteServiceName serviceName, Connection connection)
+            internal void Free(ConcurrentQueue<JsonRpcConnection> pool, JsonRpcConnection connection)
             {
                 using (_shutdownLock.DisposableRead())
                 {
-                    if (_isDisposed)
+                    if (_isDisposed || pool.Count >= _capacityPerService)
```

Comment:
is it safe to access .Count here?  i'm presuming the worse case we might dispose multiple connections unnecessarily.  but it would be good to doc what is ok here.